File: Javascript_sol/Matrix/Leetcode_74.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
// Time complexity O(m+logN) Space complexity O(1)
 var searchMatrix = function(matrix, target) {
  var x = 0
  while (x < matrix.length) {
      if (matrix[x][matrix[0].length - 1] >= target) {
          break
      };
      ++x;
  };
  if (matrix.length === x) {return false};
  
  let left = 0;
  let right = matrix[0].length;
  let mid = 0;
  
  while (left < right) {
      mid = parseInt((left + right)/2);
      if (matrix[x][mid] === target) {
          return true;
      } else if (matrix[x][mid] > target) {
          right = mid;
      } else {
          left = mid + 1;
      };
  };
  return false
};

File: Python_sol/Matrix/Leetcode_74.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # Time complexity O(m+n) Space complexity O(1)
        for i in range(len(matrix)):
            if target <= matrix[i][-1]:
                if target in matrix[i]:
                    return True
                return False

"""Better Solution
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = 0
        
        # Time complexity O(m + logN) Space complexity O(1)
        while row < len(matrix):
            if matrix[row][-1] >= target:
                break
            else:
                row += 1
        if row >= len(matrix):
            return False
        
        left = 0
        right = len(matrix[0])
        while left < right:
            mid = int((left + right)/2)
            
            if matrix[row][mid] == target:
                return True
            
            elif matrix[row][mid] > target:
                right = mid 
            
            else:
                left = mid + 1
"""

"""
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            if row[0] > target:
                break
            if row[-1] < target:
                continue

            for num in row:
                if target == num:
                    return 1

                if num > target:
                    return 0
            
            return 0
"""